File: bindings/python/jsonschema_llm_wasi.py
"""
WASI-backed wrapper for jsonschema-llm.

Uses wasmtime-py to load the universal WASI binary and exposes
convert() and rehydrate() as Python functions.

Concurrency: Each Engine owns its own Store. NOT thread-safe.
"""

import json
import os
import struct
from typing import Any, Optional

import wasmtime


_DEFAULT_WASM_PATH = os.path.join(
    os.path.dirname(__file__),
    "..", "..",
    "target", "wasm32-wasip1", "release", "jsonschema_llm_wasi.wasm",
)

JSL_RESULT_SIZE = 12  # 3 × u32 (LE)
EXPECTED_ABI_VERSION = 1
STATUS_OK = 0
STATUS_ERROR = 1


class JslError(Exception):
    """Structured error from the WASI binary."""

    def __init__(self, code: str, message: str, path: str = ""):
        self.code = code
        self.message = message
        self.path = path
        super().__init__(f"jsl error [{code}]{f' at {path}' if path else ''}: {message}")


class Engine:
    """
    WASI-backed jsonschema-llm engine.

    Usage::

        with Engine() as eng:
            result = eng.convert(schema)
            rehydrated = eng.rehydrate(data, result["codec"], schema)
    """

    def __init__(self, wasm_path: Optional[str] = None):
        path = wasm_path or os.environ.get("JSL_WASM_PATH", _DEFAULT_WASM_PATH)
        self._engine = wasmtime.Engine()
        self._module = wasmtime.Module.from_file(self._engine, path)
        self._linker = wasmtime.Linker(self._engine)
        self._linker.define_wasi()
        self._abi_verified = False

    def __enter__(self):
        return self

    def __exit__(self, *_):
        pass  # No persistent resources to clean up

    def convert(self, schema: Any, options: Optional[dict] = None) -> dict:
        """Convert a JSON Schema to LLM-compatible structured output schema."""
        schema_json = json.dumps(schema)
        # Normalize snake_case keys to kebab-case for WASI binary compatibility
        normalized = {}
        if options:
            for key, value in options.items():
                normalized[key.replace("_", "-")] = value
        opts_json = json.dumps(normalized)
        return self._call_jsl("jsl_convert", schema_json, opts_json)

    def rehydrate(self, data: Any, codec: Any, schema: Any) -> dict:
        """Rehydrate LLM output back to original schema shape."""
        data_json = json.dumps(data)
        codec_json = json.dumps(codec)
        schema_json = json.dumps(schema)
        return self._call_jsl("jsl_rehydrate", data_json, codec_json, schema_json)

    def _call_jsl(self, func_name: str, *json_args: str) -> dict:
        """Execute a WASI export following the JslResult protocol."""
        # Fresh store + instance per call (WASI modules are single-use)
        store = wasmtime.Store(self._engine)
        store.set_wasi(wasmtime.WasiConfig())
        instance = self._linker.instantiate(store, self._module)

        # ABI version handshake (once per Engine lifetime)
        if not self._abi_verified:
            exports = instance.exports(store)
            try:
                abi_fn = exports["jsl_abi_version"]
            except KeyError:
                raise RuntimeError(
                    "Incompatible WASM module: missing required 'jsl_abi_version' export"
                )
            version = abi_fn(store)
            if version != EXPECTED_ABI_VERSION:
                raise RuntimeError(
                    f"ABI version mismatch: binary={version}, expected={EXPECTED_ABI_VERSION}"
                )
            self._abi_verified = True

        memory = instance.exports(store)["memory"]
        jsl_alloc = instance.exports(store)["jsl_alloc"]
        jsl_free = instance.exports(store)["jsl_free"]
        jsl_result_free = instance.exports(store)["jsl_result_free"]
        func = instance.exports(store)[func_name]

        # Allocate and write each argument
        allocs = []
        flat_args = []
        for arg in json_args:
            data = arg.encode("utf-8")
            ptr = jsl_alloc(store, len(data))
            if ptr == 0 and len(data) > 0:
                raise RuntimeError(f"jsl_alloc returned null for {len(data)} bytes")
            memory.write(store, data, ptr)
            allocs.append((ptr, len(data)))
            flat_args.extend([ptr, len(data)])

        # Call, read, and parse — with guaranteed cleanup
        result_ptr = 0
        try:
            result_ptr = func(store, *flat_args)
            if result_ptr == 0:
                raise RuntimeError(f"{func_name} returned null result pointer")

            # Read JslResult (12 bytes: 3 × LE u32)
            result_bytes = memory.read(store, result_ptr, result_ptr + JSL_RESULT_SIZE)
            status, payload_ptr, payload_len = struct.unpack("<III", result_bytes)

            # Validate payload bounds
            mem_size = memory.data_len(store)
            if payload_ptr + payload_len > mem_size:
                raise RuntimeError(
                    f"payload out of bounds: ptr={payload_ptr} len={payload_len} memSize={mem_size}"
                )

            # Read and copy payload
            payload_bytes = memory.read(store, payload_ptr, payload_ptr + payload_len)
            payload = json.loads(payload_bytes.decode("utf-8"))

            if status == STATUS_ERROR:
                raise JslError(
                    code=payload.get("code", "unknown"),
                    message=payload.get("message", "unknown error"),
                    path=payload.get("path", ""),
                )

            return payload
        finally:
            if result_ptr != 0:
                jsl_result_free(store, result_ptr)
            for ptr, length in allocs:
                jsl_free(store, ptr, length)

File: bindings/java/src/main/java/com/jsonschema/llm/wasi/JsonSchemaLlmWasi.java
package com.jsonschema.llm.wasi;

import com.dylibso.chicory.log.SystemLogger;
import com.dylibso.chicory.runtime.ExportFunction;
import com.dylibso.chicory.runtime.HostImports;
import com.dylibso.chicory.runtime.Instance;
import com.dylibso.chicory.runtime.Memory;
import com.dylibso.chicory.runtime.Module;
import com.dylibso.chicory.wasi.WasiPreview1;
import com.dylibso.chicory.wasm.types.Value;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * WASI-backed wrapper for jsonschema-llm.
 *
 * <p>
 * Uses Chicory (pure Java, zero native deps) to load the universal WASI binary.
 *
 * <p>
 * Concurrency: Builds a new Module/Instance per call. NOT thread-safe; use one
 * JsonSchemaLlmWasi per thread or synchronize externally.
 */
public class JsonSchemaLlmWasi implements AutoCloseable {

    private static final int JSL_RESULT_SIZE = 12; // 3 × u32 (LE)
    private static final int STATUS_OK = 0;
    private static final int STATUS_ERROR = 1;
    private static final int EXPECTED_ABI_VERSION = 1;
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final File wasmFile;
    private boolean abiVerified = false;

    public JsonSchemaLlmWasi() {
        this(System.getenv("JSL_WASM_PATH") != null
                ? System.getenv("JSL_WASM_PATH")
                : "target/wasm32-wasip1/release/jsonschema_llm_wasi.wasm");
    }

    public JsonSchemaLlmWasi(String wasmPath) {
        this.wasmFile = new File(wasmPath);
    }

    @Override
    public void close() {
        // No resources to close
    }

    public JsonNode convert(Object schema) throws JslException {
        return convert(schema, null);
    }

    private String normalizeOptionsJson(Object options) throws com.fasterxml.jackson.core.JsonProcessingException {
        if (options == null)
            return "{}";
        // Serialize to JsonNode, then recursively normalize keys
        JsonNode node = MAPPER.valueToTree(options);
        if (node.isObject()) {
            node = normalizeKeys(node);
        }
        return MAPPER.writeValueAsString(node);
    }

    private JsonNode normalizeKeys(JsonNode node) {
        if (!node.isObject())
            return node;
        com.fasterxml.jackson.databind.node.ObjectNode result = MAPPER.createObjectNode();
        node.fields().forEachRemaining(entry -> {
            // Convert camelCase/snake_case to kebab-case
            String key = entry.getKey()
                    .replaceAll("([a-z])([A-Z])", "$1-$2")
                    .replace('_', '-')
                    .toLowerCase();
            result.set(key, normalizeKeys(entry.getValue()));
        });
        return result;
    }

    public JsonNode convert(Object schema, Object options) throws JslException {
        try {
            String schemaJson = MAPPER.writeValueAsString(schema);
            String optsJson = normalizeOptionsJson(options);
            return callJsl("jsl_convert", schemaJson, optsJson);
        } catch (JslException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("convert failed", e);
        }
    }

    public JsonNode rehydrate(Object data, Object codec, Object schema) throws JslException {
        try {
            String dataJson = MAPPER.writeValueAsString(data);
            String codecJson = MAPPER.writeValueAsString(codec);
            String schemaJson = MAPPER.writeValueAsString(schema);
            return callJsl("jsl_rehydrate", dataJson, codecJson, schemaJson);
        } catch (JslException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("rehydrate failed", e);
        }
    }

    JsonNode callJsl(String funcName, String... jsonArgs) throws JslException {
        // Fresh WASI + instance per call (WASI modules are single-use)
        try (WasiPreview1 wasi = new WasiPreview1(new SystemLogger())) {
            HostImports hostImports = new HostImports(wasi.toHostFunctions());
            Module module = Module.builder(wasmFile)
                    .withHostImports(hostImports)
                    .build();
            Instance instance = module.instantiate();
            Memory memory = instance.memory();

            ExportFunction jslAlloc = instance.export("jsl_alloc");
            ExportFunction jslFree = instance.export("jsl_free");
            ExportFunction jslResultFree = instance.export("jsl_result_free");
            ExportFunction func = instance.export(funcName);

            // ABI version handshake (once per engine lifetime)
            if (!abiVerified) {
                try {
                    ExportFunction abiFn = instance.export("jsl_abi_version");
                    if (abiFn == null) {
                        throw new RuntimeException(
                                "Incompatible WASM module: missing required 'jsl_abi_version' export");
                    }
                    Value[] abiResult = abiFn.apply();
                    int version = abiResult[0].asInt();
                    if (version != EXPECTED_ABI_VERSION) {
                        throw new RuntimeException(
                                "ABI version mismatch: binary=" + version + ", expected=" + EXPECTED_ABI_VERSION);
                    }
                } catch (RuntimeException e) {
                    throw e;
                } catch (Exception e) {
                    throw new RuntimeException("ABI handshake failed", e);
                }
                abiVerified = true;
            }

            // Allocate and write arguments
            List<int[]> allocs = new ArrayList<>();
            List<Value> flatArgs = new ArrayList<>();
            int resultPtr = 0;

            try {
                for (String arg : jsonArgs) {
                    byte[] bytes = arg.getBytes(StandardCharsets.UTF_8);
                    Value[] allocResult = jslAlloc.apply(Value.i32(bytes.length));
                    int ptr = allocResult[0].asInt();
                    if (ptr == 0 && bytes.length > 0) {
                        throw new RuntimeException("jsl_alloc returned null for " + bytes.length + " bytes");
                    }
                    memory.write(ptr, bytes);
                    allocs.add(new int[] { ptr, bytes.length });
                    flatArgs.add(Value.i32(ptr));
                    flatArgs.add(Value.i32(bytes.length));
                }

                // Call function
                Value[] result = func.apply(flatArgs.toArray(new Value[0]));
                resultPtr = result[0].asInt();
                if (resultPtr == 0) {
                    throw new RuntimeException(funcName + " returned null result pointer");
                }

                // Read JslResult (12 bytes: 3 × LE u32)
                byte[] resultBytes = memory.readBytes(resultPtr, JSL_RESULT_SIZE);
                ByteBuffer buf = ByteBuffer.wrap(resultBytes).order(ByteOrder.LITTLE_ENDIAN);
                int status = buf.getInt();
                int payloadPtr = buf.getInt();
                int payloadLen = buf.getInt();

                // Validate payload bounds
                if (payloadPtr < 0 || payloadLen < 0) {
                    throw new RuntimeException(
                            "invalid payload pointer/length: ptr=" + payloadPtr + " len=" + payloadLen);
                }

                // Read payload
                byte[] payloadBytes = memory.readBytes(payloadPtr, payloadLen);
                String payloadStr = new String(payloadBytes, StandardCharsets.UTF_8);

                JsonNode payload = MAPPER.readTree(payloadStr);

                if (status == STATUS_ERROR) {
                    throw new JslException(
                            payload.path("code").asText("unknown"),
                            payload.path("message").asText("unknown error"),
                            payload.path("path").asText(""));
                }

                return payload;
            } finally {
                // Always free guest memory
                if (resultPtr != 0) {
                    jslResultFree.apply(Value.i32(resultPtr));
                }
                for (int[] alloc : allocs) {
                    jslFree.apply(Value.i32(alloc[0]), Value.i32(alloc[1]));
                }
            }
        } catch (JslException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("callJsl failed: " + funcName, e);
        }
    }

    public static class JslException extends Exception {
        private final String code;
        private final String path;

        public JslException(String code, String message, String path) {
            super(String.format("jsl error [%s]%s: %s",
                    code,
                    path != null && !path.isEmpty() ? " at " + path : "",
                    message));
            this.code = code;
            this.path = path;
        }

        public String getCode() {
            return code;
        }

        public String getPath() {
            return path;
        }
    }
}
